### ISSUE TITLE
Problems with Asyncssh 1.16.0

### ISSUE THREAD

**Issue by kovaxur** [selfuryon/netdev#24]
Hi,
I'm using netdev for a few months now, thanks for the great job!

Since asyncssh 1.16.0 is out, we have problems connecting to some of our devices. It looks like, that there was a change in the part, where the known_hosts are checked. It seems, that the script for these devices just reject the connection after KEX init.

What we see on some of the cisco devices:
```
SSH2: Session disconnect – error 0x00
%SSH-4-SSH2_UNEXPECTED_MSG: Unexpected message type has arrived. Terminating the connection
```

What we see in the script output, when I enabled debugging for netdev and asyncssh:
```
2019-04-24 11:03:55,303 [conn=284] Requesting key exchange
2019-04-24 11:03:55,303 [conn=284]   Key exchange algs: curve25519-sha256,curve25519-sha256@libssh.org,curve448-sha512,ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,ecdh-sha2-1.3.132.0.10,diffie-hellman-group-exchange-sha256,di
ffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1,diffie-hellman-group14-sha1,diffie-hellman-group14-sha256,diffie-hellman-group15-sha512,diffie-hellman-group16-sha512,diffie-hellman-group17-sha512,diffie-hellman-group18-sha512,
rsa2048-sha256,rsa1024-sha1
2019-04-24 11:03:55,303 [conn=284]   Host key algs: ssh-ed25519-cert-v01@openssh.com,ssh-ed448-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp256-cert-v01@openssh.com
,ecdsa-sha2-1.3.132.0.10-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ssh-dss-cert-v01@openssh.com,ssh-ed25519,ssh-ed448,ecdsa-sha2-nistp521,ecdsa-sha2-nistp384,ecdsa-sha2-nistp256,ecdsa-sha2-1.3.132.0.10,rsa-sha2-256,rsa-sha2-512,ss
h-rsa,ssh-dss
2019-04-24 11:03:55,303 [conn=284]   Encryption algs: aes256-gcm@openssh.com,aes128-gcm@openssh.com,aes256-ctr,aes192-ctr,aes128-ctr,aes256-cbc,aes192-cbc,aes128-cbc,3des-cbc,blowfish-cbc,cast128-cbc,arcfour256,arcfour128,arcfour
2019-04-24 11:03:55,304 [conn=284]   MAC algs: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,hmac-md5-etm@openssh.com,hmac-sha2-256-96-etm@openssh.co
m,hmac-sha2-512-96-etm@openssh.com,hmac-sha1-96-etm@openssh.com,hmac-md5-96-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1,hmac-md5,hmac-sha2-256-96,hmac-sha2-512-96,hmac-sha1-96,hmac-md5-96
2019-04-24 11:03:55,304 [conn=284]   Compression algs: zlib@openssh.com,zlib,none
2019-04-24 11:03:55,306 [conn=284, pktid=0] Sent MSG_KEXINIT (20), 2217 bytes
  00000000: 14 c8 bd 23 97 62 5b 90 7a 77 c3 b8 f0 a7 7f 54  ...#.b[.zw...T
  00000010: 65 00 00 01 ca 63 75 72 76 65 32 35 35 31 39 2d  e....curve25519-
  00000020: 73 68 61 32 35 36 2c 63 75 72 76 65 32 35 35 31  sha256,curve2551
  00000030: 39 2d 73 68 61 32 35 36 40 6c 69 62 73 73 68 2e  9-sha256@libssh.
  00000040: 6f 72 67 2c 63 75 72 76 65 34 34 38 2d 73 68 61  org,curve448-sha
  00000050: 35 31 32 2c 65 63 64 68 2d 73 68 61 32 2d 6e 69  512,ecdh-sha2-ni
  00000060: 73 74 70 35 32 31 2c 65 63 64 68 2d 73 68 61 32  stp521,ecdh-sha2
  00000070: 2d 6e 69 73 74 70 33 38 34 2c 65 63 64 68 2d 73  -nistp384,ecdh-s
  00000080: 68 61 32 2d 6e 69 73 74 70 32 35 36 2c 65 63 64  ha2-nistp256,ecd
  00000090: 68 2d 73 68 61 32 2d 31 2e 33 2e 31 33 32 2e 30  h-sha2-1.3.132.0
  000000a0: 2e 31 30 2c 64 69 66 66 69 65 2d 68 65 6c 6c 6d  .10,diffie-hellm
  000000b0: 61 6e 2d 67 72 6f 75 70 2d 65 78 63 68 61 6e 67  an-group-exchang
  000000c0: 65 2d 73 68 61 32 35 36 2c 64 69 66 66 69 65 2d  e-sha256,diffie-
  000000d0: 68 65 6c 6c 6d 61 6e 2d 67 72 6f 75 70 2d 65 78  hellman-group-ex
  000000e0: 63 68 61 6e 67 65 2d 73 68 61 31 2c 64 69 66 66  change-sha1,diff
  000000f0: 69 65 2d 68 65 6c 6c 6d 61 6e 2d 67 72 6f 75 70  ie-hellman-group
  00000100: 31 2d 73 68 61 31 2c 64 69 66 66 69 65 2d 68 65  1-sha1,diffie-he
  00000110: 6c 6c 6d 61 6e 2d 67 72 6f 75 70 31 34 2d 73 68  llman-group14-sh
  00000120: 61 31 2c 64 69 66 66 69 65 2d 68 65 6c 6c 6d 61  a1,diffie-hellma
  00000130: 6e 2d 67 72 6f 75 70 31 34 2d 73 68 61 32 35 36  n-group14-sha256
  00000140: 2c 64 69 66 66 69 65 2d 68 65 6c 6c 6d 61 6e 2d  ,diffie-hellman-
  00000150: 67 72 6f 75 70 31 35 2d 73 68 61 35 31 32 2c 64  group15-sha512,d
  00000160: 69 66 66 69 65 2d 68 65 6c 6c 6d 61 6e 2d 67 72  iffie-hellman-gr
  00000170: 6f 75 70 31 36 2d 73 68 61 35 31 32 2c 64 69 66  oup16-sha512,dif
  00000180: 66 69 65 2d 68 65 6c 6c 6d 61 6e 2d 67 72 6f 75  fie-hellman-grou
  00000190: 70 31 37 2d 73 68 61 35 31 32 2c 64 69 66 66 69  p17-sha512,diffi
  000001a0: 65 2d 68 65 6c 6c 6d 61 6e 2d 67 72 6f 75 70 31  e-hellman-group1
  000001b0: 38 2d 73 68 61 35 31 32 2c 72 73 61 32 30 34 38  8-sha512,rsa2048
  000001c0: 2d 73 68 61 32 35 36 2c 72 73 61 31 30 32 34 2d  -sha256,rsa1024-
  000001d0: 73 68 61 31 2c 65 78 74 2d 69 6e 66 6f 2d 63 00  sha1,ext-info-c.
  000001e0: 00 01 b5 73 73 68 2d 65 64 32 35 35 31 39 2d 63  ...ssh-ed25519-c
  000001f0: 65 72 74 2d 76 30 31 40 6f 70 65 6e 73 73 68 2e  ert-v01@openssh.
  00000200: 63 6f 6d 2c 73 73 68 2d 65 64 34 34 38 2d 63 65  com,ssh-ed448-ce
  00000210: 72 74 2d 76 30 31 40 6f 70 65 6e 73 73 68 2e 63  rt-v01@openssh.c
  00000220: 6f 6d 2c 65 63 64 73 61 2d 73 68 61 32 2d 6e 69  om,ecdsa-sha2-ni
  00000230: 73 74 70 35 32 31 2d 63 65 72 74 2d 76 30 31 40  stp521-cert-v01@
  00000240: 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c 65 63 64 73  openssh.com,ecds
  00000250: 61 2d 73 68 61 32 2d 6e 69 73 74 70 33 38 34 2d  a-sha2-nistp384-
  00000260: 63 65 72 74 2d 76 30 31 40 6f 70 65 6e 73 73 68  cert-v01@openssh
  00000270: 2e 63 6f 6d 2c 65 63 64 73 61 2d 73 68 61 32 2d  .com,ecdsa-sha2-
  00000280: 6e 69 73 74 70 32 35 36 2d 63 65 72 74 2d 76 30  nistp256-cert-v0
  00000290: 31 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c 65 63  1@openssh.com,ec
  000002a0: 64 73 61 2d 73 68 61 32 2d 31 2e 33 2e 31 33 32  dsa-sha2-1.3.132
  000002b0: 2e 30 2e 31 30 2d 63 65 72 74 2d 76 30 31 40 6f  .0.10-cert-v01@o
  000002c0: 70 65 6e 73 73 68 2e 63 6f 6d 2c 73 73 68 2d 72  penssh.com,ssh-r
  000002d0: 73 61 2d 63 65 72 74 2d 76 30 31 40 6f 70 65 6e  sa-cert-v01@open
  000002e0: 73 73 68 2e 63 6f 6d 2c 73 73 68 2d 64 73 73 2d  ssh.com,ssh-dss-
  000002f0: 63 65 72 74 2d 76 30 31 40 6f 70 65 6e 73 73 68  cert-v01@openssh
  00000300: 2e 63 6f 6d 2c 73 73 68 2d 65 64 32 35 35 31 39  .com,ssh-ed25519
  00000310: 2c 73 73 68 2d 65 64 34 34 38 2c 65 63 64 73 61  ,ssh-ed448,ecdsa
  00000320: 2d 73 68 61 32 2d 6e 69 73 74 70 35 32 31 2c 65  -sha2-nistp521,e
  00000330: 63 64 73 61 2d 73 68 61 32 2d 6e 69 73 74 70 33  cdsa-sha2-nistp3
  00000340: 38 34 2c 65 63 64 73 61 2d 73 68 61 32 2d 6e 69  84,ecdsa-sha2-ni
  00000350: 73 74 70 32 35 36 2c 65 63 64 73 61 2d 73 68 61  stp256,ecdsa-sha
  00000360: 32 2d 31 2e 33 2e 31 33 32 2e 30 2e 31 30 2c 72  2-1.3.132.0.10,r
  00000370: 73 61 2d 73 68 61 32 2d 32 35 36 2c 72 73 61 2d  sa-sha2-256,rsa-
  00000380: 73 68 61 32 2d 35 31 32 2c 73 73 68 2d 72 73 61  sha2-512,ssh-rsa
  00000390: 2c 73 73 68 2d 64 73 73 00 00 00 af 61 65 73 32  ,ssh-dss....aes2
  000003a0: 35 36 2d 67 63 6d 40 6f 70 65 6e 73 73 68 2e 63  56-gcm@openssh.c
  000003b0: 6f 6d 2c 61 65 73 31 32 38 2d 67 63 6d 40 6f 70  om,aes128-gcm@op
  000003c0: 65 6e 73 73 68 2e 63 6f 6d 2c 61 65 73 32 35 36  enssh.com,aes256
  000003d0: 2d 63 74 72 2c 61 65 73 31 39 32 2d 63 74 72 2c  -ctr,aes192-ctr,
  000003e0: 61 65 73 31 32 38 2d 63 74 72 2c 61 65 73 32 35  aes128-ctr,aes25
  000003f0: 36 2d 63 62 63 2c 61 65 73 31 39 32 2d 63 62 63  6-cbc,aes192-cbc
  00000400: 2c 61 65 73 31 32 38 2d 63 62 63 2c 33 64 65 73  ,aes128-cbc,3des
  00000410: 2d 63 62 63 2c 62 6c 6f 77 66 69 73 68 2d 63 62  -cbc,blowfish-cb
  00000420: 63 2c 63 61 73 74 31 32 38 2d 63 62 63 2c 61 72  c,cast128-cbc,ar
  00000430: 63 66 6f 75 72 32 35 36 2c 61 72 63 66 6f 75 72  cfour256,arcfour
  00000440: 31 32 38 2c 61 72 63 66 6f 75 72 00 00 00 af 61  128,arcfour....a
  00000450: 65 73 32 35 36 2d 67 63 6d 40 6f 70 65 6e 73 73  es256-gcm@openss
  00000460: 68 2e 63 6f 6d 2c 61 65 73 31 32 38 2d 67 63 6d  h.com,aes128-gcm
  00000470: 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c 61 65 73  @openssh.com,aes
  00000480: 32 35 36 2d 63 74 72 2c 61 65 73 31 39 32 2d 63  256-ctr,aes192-c
  00000490: 74 72 2c 61 65 73 31 32 38 2d 63 74 72 2c 61 65  tr,aes128-ctr,ae
  000004a0: 73 32 35 36 2d 63 62 63 2c 61 65 73 31 39 32 2d  s256-cbc,aes192-
  000004b0: 63 62 63 2c 61 65 73 31 32 38 2d 63 62 63 2c 33  cbc,aes128-cbc,3
  000004c0: 64 65 73 2d 63 62 63 2c 62 6c 6f 77 66 69 73 68  des-cbc,blowfish
  000004d0: 2d 63 62 63 2c 63 61 73 74 31 32 38 2d 63 62 63  -cbc,cast128-cbc
  000004e0: 2c 61 72 63 66 6f 75 72 32 35 36 2c 61 72 63 66  ,arcfour256,arcf
  000004f0: 6f 75 72 31 32 38 2c 61 72 63 66 6f 75 72 00 00  our128,arcfour..
  00000500: 01 ad 75 6d 61 63 2d 36 34 2d 65 74 6d 40 6f 70  ..umac-64-etm@op
  00000510: 65 6e 73 73 68 2e 63 6f 6d 2c 75 6d 61 63 2d 31  enssh.com,umac-1
  00000520: 32 38 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63  28-etm@openssh.c
  00000530: 6f 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 32 35 36  om,hmac-sha2-256
  00000540: 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f 6d  -etm@openssh.com
  00000550: 2c 68 6d 61 63 2d 73 68 61 32 2d 35 31 32 2d 65  ,hmac-sha2-512-e
  00000560: 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c 68  tm@openssh.com,h
  00000570: 6d 61 63 2d 73 68 61 31 2d 65 74 6d 40 6f 70 65  mac-sha1-etm@ope
  00000580: 6e 73 73 68 2e 63 6f 6d 2c 68 6d 61 63 2d 6d 64  nssh.com,hmac-md
  00000590: 35 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f  5-etm@openssh.co
  000005a0: 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 32 35 36 2d  m,hmac-sha2-256-
  000005b0: 39 36 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63  96-etm@openssh.c
  000005c0: 6f 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 35 31 32  om,hmac-sha2-512
  000005d0: 2d 39 36 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e  -96-etm@openssh.
  000005e0: 63 6f 6d 2c 68 6d 61 63 2d 73 68 61 31 2d 39 36  com,hmac-sha1-96
  000005f0: 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f 6d  -etm@openssh.com
  00000600: 2c 68 6d 61 63 2d 6d 64 35 2d 39 36 2d 65 74 6d  ,hmac-md5-96-etm
  00000610: 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c 75 6d 61  @openssh.com,uma
  00000620: 63 2d 36 34 40 6f 70 65 6e 73 73 68 2e 63 6f 6d  c-64@openssh.com
  00000630: 2c 75 6d 61 63 2d 31 32 38 40 6f 70 65 6e 73 73  ,umac-128@openss
  00000640: 68 2e 63 6f 6d 2c 68 6d 61 63 2d 73 68 61 32 2d  h.com,hmac-sha2-
  00000650: 32 35 36 2c 68 6d 61 63 2d 73 68 61 32 2d 35 31  256,hmac-sha2-51
  00000660: 32 2c 68 6d 61 63 2d 73 68 61 31 2c 68 6d 61 63  2,hmac-sha1,hmac
  00000670: 2d 6d 64 35 2c 68 6d 61 63 2d 73 68 61 32 2d 32  -md5,hmac-sha2-2
  00000680: 35 36 2d 39 36 2c 68 6d 61 63 2d 73 68 61 32 2d  56-96,hmac-sha2-
  00000690: 35 31 32 2d 39 36 2c 68 6d 61 63 2d 73 68 61 31  512-96,hmac-sha1
  000006a0: 2d 39 36 2c 68 6d 61 63 2d 6d 64 35 2d 39 36 00  -96,hmac-md5-96.
  000006b0: 00 01 ad 75 6d 61 63 2d 36 34 2d 65 74 6d 40 6f  ...umac-64-etm@o
  000006c0: 70 65 6e 73 73 68 2e 63 6f 6d 2c 75 6d 61 63 2d  penssh.com,umac-
  000006d0: 31 32 38 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e  128-etm@openssh.
  000006e0: 63 6f 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 32 35  com,hmac-sha2-25
  000006f0: 36 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f  6-etm@openssh.co
  00000700: 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 35 31 32 2d  m,hmac-sha2-512-
  00000710: 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c  etm@openssh.com,
  00000720: 68 6d 61 63 2d 73 68 61 31 2d 65 74 6d 40 6f 70  hmac-sha1-etm@op
  00000730: 65 6e 73 73 68 2e 63 6f 6d 2c 68 6d 61 63 2d 6d  enssh.com,hmac-m
  00000740: 64 35 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63  d5-etm@openssh.c
  00000750: 6f 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 32 35 36  om,hmac-sha2-256
  00000760: 2d 39 36 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e  -96-etm@openssh.
  00000770: 63 6f 6d 2c 68 6d 61 63 2d 73 68 61 32 2d 35 31  com,hmac-sha2-51
  00000780: 32 2d 39 36 2d 65 74 6d 40 6f 70 65 6e 73 73 68  2-96-etm@openssh
  00000790: 2e 63 6f 6d 2c 68 6d 61 63 2d 73 68 61 31 2d 39  .com,hmac-sha1-9
  000007a0: 36 2d 65 74 6d 40 6f 70 65 6e 73 73 68 2e 63 6f  6-etm@openssh.co
  000007b0: 6d 2c 68 6d 61 63 2d 6d 64 35 2d 39 36 2d 65 74  m,hmac-md5-96-et
  000007c0: 6d 40 6f 70 65 6e 73 73 68 2e 63 6f 6d 2c 75 6d  m@openssh.com,um
  000007d0: 61 63 2d 36 34 40 6f 70 65 6e 73 73 68 2e 63 6f  ac-64@openssh.co
  000007e0: 6d 2c 75 6d 61 63 2d 31 32 38 40 6f 70 65 6e 73  m,umac-128@opens
  000007f0: 73 68 2e 63 6f 6d 2c 68 6d 61 63 2d 73 68 61 32  sh.com,hmac-sha2
  00000800: 2d 32 35 36 2c 68 6d 61 63 2d 73 68 61 32 2d 35  -256,hmac-sha2-5
  00000810: 31 32 2c 68 6d 61 63 2d 73 68 61 31 2c 68 6d 61  12,hmac-sha1,hma
  00000820: 63 2d 6d 64 35 2c 68 6d 61 63 2d 73 68 61 32 2d  c-md5,hmac-sha2-
  00000830: 32 35 36 2d 39 36 2c 68 6d 61 63 2d 73 68 61 32  256-96,hmac-sha2
  00000840: 2d 35 31 32 2d 39 36 2c 68 6d 61 63 2d 73 68 61  -512-96,hmac-sha
  00000850: 31 2d 39 36 2c 68 6d 61 63 2d 6d 64 35 2d 39 36  1-96,hmac-md5-96
  00000860: 00 00 00 1a 7a 6c 69 62 40 6f 70 65 6e 73 73 68  ....zlib@openssh
  00000870: 2e 63 6f 6d 2c 7a 6c 69 62 2c 6e 6f 6e 65 00 00  .com,zlib,none..
  00000880: 00 1a 7a 6c 69 62 40 6f 70 65 6e 73 73 68 2e 63  ..zlib@openssh.c
  00000890: 6f 6d 2c 7a 6c 69 62 2c 6e 6f 6e 65 00 00 00 00  om,zlib,none....
  000008a0: 00 00 00 00 00 00 00 00 00                       .........
2019-04-24 11:03:55,306 [conn=284, pktid=0] Received MSG_KEXINIT (20), 271 bytes
  00000000: 14 78 4c 9d 24 47 05 21 54 3b 6f bc 59 c2 1d 70  .xL.$G.!T;o.Y..p
  00000010: 1e 00 00 00 1a 64 69 66 66 69 65 2d 68 65 6c 6c  .....diffie-hell
  00000020: 6d 61 6e 2d 67 72 6f 75 70 31 2d 73 68 61 31 00  man-group1-sha1.
  00000030: 00 00 07 73 73 68 2d 72 73 61 00 00 00 29 61 65  ...ssh-rsa...)ae
  00000040: 73 31 32 38 2d 63 62 63 2c 33 64 65 73 2d 63 62  s128-cbc,3des-cb
  00000050: 63 2c 61 65 73 31 39 32 2d 63 62 63 2c 61 65 73  c,aes192-cbc,aes
  00000060: 32 35 36 2d 63 62 63 00 00 00 29 61 65 73 31 32  256-cbc...)aes12
  00000070: 38 2d 63 62 63 2c 33 64 65 73 2d 63 62 63 2c 61  8-cbc,3des-cbc,a
  00000080: 65 73 31 39 32 2d 63 62 63 2c 61 65 73 32 35 36  es192-cbc,aes256
  00000090: 2d 63 62 63 00 00 00 2b 68 6d 61 63 2d 73 68 61  -cbc...+hmac-sha
  000000a0: 31 2c 68 6d 61 63 2d 73 68 61 31 2d 39 36 2c 68  1,hmac-sha1-96,h
  000000b0: 6d 61 63 2d 6d 64 35 2c 68 6d 61 63 2d 6d 64 35  mac-md5,hmac-md5
  000000c0: 2d 39 36 00 00 00 2b 68 6d 61 63 2d 73 68 61 31  -96...+hmac-sha1
  000000d0: 2c 68 6d 61 63 2d 73 68 61 31 2d 39 36 2c 68 6d  ,hmac-sha1-96,hm
  000000e0: 61 63 2d 6d 64 35 2c 68 6d 61 63 2d 6d 64 35 2d  ac-md5,hmac-md5-
  000000f0: 39 36 00 00 00 04 6e 6f 6e 65 00 00 00 04 6e 6f  96....none....no
  00000100: 6e 65 00 00 00 00 00 00 00 00 00 00 00 00 00     ne.............
2019-04-24 11:03:55,306 [conn=284] Received key exchange request
2019-04-24 11:03:55,306 [conn=284]   Key exchange algs: diffie-hellman-group1-sha1
2019-04-24 11:03:55,306 [conn=284]   Host key algs: ssh-rsa
2019-04-24 11:03:55,306 [conn=284]   Client to server:
2019-04-24 11:03:55,306 [conn=284]     Encryption algs: aes128-cbc,3des-cbc,aes192-cbc,aes256-cbc
2019-04-24 11:03:55,306 [conn=284]     MAC algs: hmac-sha1,hmac-sha1-96,hmac-md5,hmac-md5-96
2019-04-24 11:03:55,306 [conn=284]     Compression algs: none
2019-04-24 11:03:55,306 [conn=284]   Server to client:
2019-04-24 11:03:55,307 [conn=284]     Encryption algs: aes128-cbc,3des-cbc,aes192-cbc,aes256-cbc
2019-04-24 11:03:55,307 [conn=284]     MAC algs: hmac-sha1,hmac-sha1-96,hmac-md5,hmac-md5-96
2019-04-24 11:03:55,307 [conn=284]     Compression algs: none
2019-04-24 11:03:55,307 [conn=284] Beginning key exchange
2019-04-24 11:03:55,307 [conn=284]   Key exchange alg: diffie-hellman-group1-sha1
2019-04-24 11:03:55,311 [conn=284, pktid=1] Sent MSG_KEXDH_INIT (30), 133 bytes
  00000000: 1e 00 00 00 80 3e 65 5c af d2 40 65 d0 dc 07 ff  .....>e\..@e....
  00000010: 91 cc 82 6f 47 02 46 67 92 98 ae 3b ae 64 16 d4  ...oG.Fg...;.d..
  00000020: a5 6d 9a e2 a0 c7 43 ff c9 b9 32 f3 eb 7c 02 8d  .m....C...2..|..
  00000030: 02 6e bb 2d e1 32 45 7e 3a 68 ab 47 c7 14 1b b4  .n.-.2E~:h.G....
  00000040: 20 6a 61 45 95 f9 35 61 8a 25 96 5e 76 23 e3 46   jaE..5a.%.^v#.F
  00000050: 5f 57 10 5e 27 63 ef f1 ea 30 84 07 ef 91 ce c4  _W.^'c...0......
  00000060: 86 ba e1 57 6a 30 2b 33 c1 1e 66 83 98 72 f3 ed  ...Wj0+3..f..r..
  00000070: 2b 7d 05 54 cf af 2e 1f 96 d7 29 92 62 87 2a a6  +}.T......).b.*.
  00000080: 25 e8 b2 06 7c                                   %...|
2019-04-24 11:03:55,746 [conn=284] Connection failure: [Errno 104] Connection reset by peer.
```

Temp solution was to fix the asyncssh version: asyncssh==1.15.1

**Comment by selfuryon** [comment]
Hello!
Thanks for using my library!
It's an interesting issue, I'll take a deep investigation to it tomorrow and write down my thoughts about it here.

**Comment by selfuryon** [comment]
Hello! Can you also write with which type of devices you find this problem?

**Comment by kovaxur** [comment]
Hi,
We have around 500 devices, but only ~30 had problems, there are switches, routers and firewalls in the list, for example:
WS-C2960-24TT-L - 12.2(44)SE6
WS-C3560-24PS - 12.2(25)SED1
WS-C3560V2-24PS - 12.2(55)SE5
ASA 5515-X - 9.6(4)8
ASA 5515-X -  9.1(7)23
WS-C3750X-24 - 12.2(55)SE5
C2811 - 12.4(5b)

It looks like that devices with older IOS versions are affected, but there are some firewalls in the list with relatively new OS versions.

**Comment by selfuryon** [comment]
Hello!
I can't reproduce this problem =\
Can you add some more information about that:
- Is the issue always persistent for these devices?
- Do you have the issue even if you are connecting only to only one device from this list?
- Do you have also the issue with asyncssh 1.16.1?

**Comment by selfuryon** [comment]
Another way - did you try to change ssh public/private key for authentication?

**Comment by kovaxur** [comment]
Hi,
Yes, I tried with different keys, but same results.

I will check with 1.16.1 next week. Then I think you can close this, as it's probably about asyncssh, but still a good-to-know if someone has the same issues.

**Comment by kovaxur** [comment]
After doing some more troubleshooting, it turned out, that some old devices support only insecure kex and encryption algorithms. 
These must be allowed manually via the netdev.create(kex_algs=, encryption_alges=) function
Supported algorithms:
- https://asyncssh.readthedocs.io/en/latest/api.html#kexalgs
- https://asyncssh.readthedocs.io/en/latest/api.html#encryptionalgs

**Comment by kovaxur** [comment]
Okay, what  I found was just some side issue, I believe I found the real. 

So the problem was, that when the client sent out the **key exchange init** message to the server (telco device), it just responded with a reset, nothing else.

SSH was working from the Linux box directly but this error happened from the script. So I did some packet capture to see, what's the difference between them, and after a few hours I spotted the difference in the signature algorithms. 

In netdev, it's not possible to set the **server_host_key_algs** algorithm, but after patching it, I set only to "ssh-rsa" and now it started to work.
https://asyncssh.readthedocs.io/en/latest/api.html#publickeyalgs

So we should add a new argument to the connect_params_dict: server_host_key_algs (and also to init function)
